drivers/include/sm_pwm_01c.h

```diff
@@ -0,0 +1,199 @@
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_sm_pwm_01c SM_PWM_01C dust sensor
+ * @ingroup     drivers_sensors
+ * @brief       Driver for Amphenol SM_PWM_01C infrared dust sensor
+ * @{
+ *
+ *
+ * * About
+ * =====
+ *
+ * This driver provides an interface for the Amphenol SM-PWM-Sensor.
+ * The Datasheet can be found [here](https://www.cdiweb.com/datasheets/telaire-amphenol/01c%20dust%20sensor%20datasheet.pdf).
+ * and the more complete application note [here](https://www.sgbotic.com/products/datasheets/sensors/app-SM-PWM-01C.pdf)
+ *
+ * The device can measure small particles (1~ 2μm) and large particle (3 ~10μm),
+ * so similar to PM2.5 and PM10. The dust sensor cannot count particles only
+ * measure estimated concentrations.
+ *
+ * It is recommended to compute values over a 30s moving average. By default
+ * a moving average is used since the module MODULE_SM_PWM_01C_MA is
+ * activated by default. To save memory an exponential average can be used
+ * by disabling this module.
+ *
+ * @file
+ * @brief       SM_PWM_01C Device Driver
+ *
+ * @author      Francisco Molina <francois-xavier.molina@inria.fr>
+ */
+
+#ifndef SM_PWM_01C_H
+#define SM_PWM_01C_H
+
+#include <inttypes.h>
+
+#include "timex.h"
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup    drivers_sm_pwm_01c_conf  SM_PWM_01C  compile configurations
+ * @ingroup     drivers_sm_pwm_01c
+ * @ingroup     config
+ * @{
+ */
+
+/**
+ * @def     CONFIG_SM_PWM_01C_SAMPLE_TIME
+ *
+ * @brief   Frequency at witch LPO % is calculated
+ */
+#ifndef CONFIG_SM_PWM_01C_SAMPLE_TIME
+#define CONFIG_SM_PWM_01C_SAMPLE_TIME           (100*US_PER_MS)
+#endif
+
+/**
+ * @def     CONFIG_SM_PWM_01C_WINDOW_TIME
+ *
+ * @brief   Length in time of the measuring window, recommended 5-30s
+ */
+#ifndef CONFIG_SM_PWM_01C_WINDOW_TIME
+#define CONFIG_SM_PWM_01C_WINDOW_TIME     (10*US_PER_SEC)
+#endif
+
+#if defined(MODULE_SM_PWM_01C_MA) || defined(DOXYGEN)
+/*
+ * @brief   Length in time of the measuring window
+ */
+#define SM_PWM_01C_BUFFER_LEN    (CONFIG_SM_PWM_01C_WINDOW_TIME / \
+                                  CONFIG_SM_PWM_01C_SAMPLE_TIME)
+#else
+
+/**
+ * @def     CONFIG_SM_PWM_01C_EXP_WEIGHT
+ *
+ * @brief   Weight of the exponential average filter where:
+ *          CONFIG_SM_PWM_01C_EXP_WEIGHT = 1 / (1 - alpha).
+ *
+ * @note    Should be chosen wisely, it can be done my minimizing MSE
+ *          or other algorithms as Marquardt procedure.
+ */
+#ifndef CONFIG_SM_PWM_01C_EXP_WEIGHT
+#define CONFIG_SM_PWM_01C_EXP_WEIGHT            (5)
+#endif
+#endif
+
+/** @} */
+
+#if defined(MODULE_SM_PWM_01C_MA) || defined(DOXYGEN)
+/**
+ * @brief   Circular buffer holding moving average values
+ * @internal
+ *
+ */
+typedef struct {
+    uint16_t buf[SM_PWM_01C_BUFFER_LEN]; /**< circular buffer memory */
+    size_t head;                         /**< current buffer head */
+} circ_buf_t;
```

Comment:
Would it make sense to use http://doc.riot-os.org/cib_8h.html instead?

Reply:
I didn't use it because the cib will note over-write the buffer content and here I do want to do so(I think since I made this choice a while ago :)).